group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupRegexManager.java

```diff
@@ -0,0 +1,388 @@
+            } catch (Exception e) {
+                log.error("Error while evaluating group regex", e);
+            }
+        }
+        log.debug("Stopping regex resolution thread");
+    }
+
+    /**
+     * Evaluate the given regular expression if it's not resolved yet.
+     *
+     * @return True if the regular expression is resolved as part of this execution. False if it was already resolved.
+     */
+    private Resolution maybeEvalRegex(
+        RegexKey groupRegexKey
+    ) {
+        String groupId = groupRegexKey.groupId();
+        Pattern pattern = groupRegexKey.pattern();
+        if (isResolved(groupId, pattern)) {
+            return null;
+        }
+        long start = System.currentTimeMillis();
+        Set<String> allTopics = this.metadataImage.topics().topicsByName().keySet();
+
+        Set<String> matchingTopics = new HashSet<>();
+        for (String topic : allTopics) {
+            if (pattern.matcher(topic).matches()) {
+                matchingTopics.add(topic);
+            }
+        }
+        log.info("Completed evaluating regex {} in {} ms against {} topics. Matching topics found: {}",
+            pattern, System.currentTimeMillis() - start, allTopics.size(), matchingTopics);
+
+        return new Resolution.Builder()
+            .withMatchingTopics(matchingTopics)
+            .withMetadataVersion(0) // TODO: integrate metadata version
+            .build();
+    }
+
+    /**
+     * Trigger write operation to persist the resolution of the regex.
+     *
+     * @param key The identifier for the regular expression (groupId and the regular expression).
+     * @param resolution The result of the evaluation of the regular expression.
+     */
+    private void writeResolvedRegex(RegexKey key, Resolution resolution) {
+        // TODO: piggybacking on the timer.schedule as initial approach but consider skipping it and trigger write
+        //  operation directly.
+        timer.schedule(
+            regexEvalAttemptKey(key.groupId(), key.pattern().toString(), resolution.metadataVersion()),
+            0,
+            TimeUnit.MILLISECONDS,
+            false,
+            () -> consumerGroupRegexResolutionCompleted(key, resolution)
+        );
+    }
+
+    /**
+     * Generate an updated record to persist the regular expression with its matching topics. All regular
+     * expressions for a single group are stored in the same record.
+     *
+     * @return The CoordinatorResult to be applied.
+     * @param <T> The type of the CoordinatorResult.
+     */
+    private <T> CoordinatorResult<T, CoordinatorRecord> consumerGroupRegexResolutionCompleted(
+        RegexKey key,
+        Resolution resolvedRegex
+    ) {
+        log.debug("Generating record with newly resolved regex {} for group {}", key.pattern(), key.groupId());
+        CoordinatorRecord r = GroupCoordinatorRecordHelpers.newConsumerGroupRegexRecord(key, resolvedRegex);
+        return new CoordinatorResult<>(Collections.singletonList(r));
```

Comment:
We need to add some validation here because we don't want to generate a record if it is not necessary. For instance, we must ensure that the group exists. We must ensure that the regex is still used by the group. We could also check whether the resolved topic names have changed. This is important because without those we may generate a record that will never be deleted.

We also need to trigger a rebalance by bumping the group epoch if the resolved topic names have changed and we must re-compute the subscription metadata because the subscription metadata must cover all the subscribed topics regardless from where they come. We could also compute the new assignment or we could rely on a subsequent heartbeat to do it.

Thinking about this, I wonder if we should rather have this logic in the GroupMetadataManager in order to keep all the metadata bookkeeping centralized. One alternative that we could consider is to turn this manager into a basic executor that other managers could use to offload heavy tasks. It would be a construct similar to the Timer.

```
executor.schedule(
    () -> {
        // Resolve the regex
    },
    (result, error) -> {
        // Callback with the result
        // Validate & generate records
    }
);
```

I suppose that we would still need a way to de-duplicate tasks based on a key like you did. They are other ways to design it. What do you think?

Reply:
I do like the executor approach, the regexManger is indeed doing that at the moment: responsible for executing an async operation (eval regex), and performing an action when done (writing records). Makes sense as an executor with callback as you suggested. Only added a TODO to follow-up with the refactor